subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -85,34 +85,58 @@ private boolean isLeafTest(TestIdentifier identifier) {
 
     @Override
     public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-        if (isLeafTestOrFailedContainer(testIdentifier, testExecutionResult)) {
-            if (!isLeafTest(testIdentifier)) {
-                // only leaf methods triggered start events previously
-                // so here we need to add the missing start events
-                adapter.testStarted(testIdentifier.getUniqueId(), getDescriptor(testIdentifier));
-            }
-            switch (testExecutionResult.getStatus()) {
-                case SUCCESSFUL:
-                    adapter.testFinished(testIdentifier.getUniqueId());
-                    break;
-                case FAILED:
-                    adapter.testFailure(testIdentifier.getUniqueId(), getDescriptor(testIdentifier), testExecutionResult.getThrowable().get());
-                    adapter.testFinished(testIdentifier.getUniqueId());
-                    break;
-                case ABORTED:
-                    adapter.testAssumptionFailure(testIdentifier.getUniqueId());
-                    adapter.testFinished(testIdentifier.getUniqueId());
-                    break;
-                default:
-                    throw new AssertionError("Invalid Status: " + testExecutionResult.getStatus());
-            }
+        if (testFailedBeforeTestClassStart(testIdentifier, testExecutionResult)) {
+            executionFailedBeforeTestClassStart(testIdentifier, testExecutionResult);
+        } else {
+            executionFinishedAfterTestClassStart(testIdentifier, testExecutionResult);
+        }
+    }
+
+    private void executionFailedBeforeTestClassStart(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+        reportTestClassStarted("Initializing");
```

Comment:
❌ Why do we synthesize a test class here? Don't we know which class it tried to run? If there is a failure before we even start executing classes, we should probably have a dedicated method on our listener instead of faking a class event.

Reply:
Without this fake class event, [this assertion](https://github.com/gradle/gradle/blob/f24146cc5f956a1b03a515bbee4225e6f861b6ab/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/results/AttachParentTestResultProcessor.java#L33) will fail and break the test event infrastructure. At this point, Jupiter engine is being initialized so we don't know the class name. Instead of tweaking event infrastructure, I think a fake event here is simple and cheap, because this fake class is only used for:

- Display in the console progress bar "Executing class XXXX"
- Avoid the above failed assertion

It's not for rendering report.

Reply:
>  Instead of tweaking event infrastructure, I think a fake event here is simple and cheap

We should handle initialization errors explicitly instead of pretending that a class was executed. Let's not add workarounds here, we are in control of our own infrastructure.